src/datasets/packaged_modules/imagefolder/imagefolder.py

```diff
@@ -0,0 +1,194 @@
+import os
+from dataclasses import dataclass
+from typing import List, Optional
+
+import datasets
+from datasets.tasks import ImageClassification
+
+
+logger = datasets.utils.logging.get_logger(__name__)
+
+
+@dataclass
+class ImageFolderConfig(datasets.BuilderConfig):
+    """BuilderConfig for ImageFolder."""
+
+    features: Optional[datasets.Features] = None
+
+
+class ImageFolder(datasets.GeneratorBasedBuilder):
+    BUILDER_CONFIG_CLASS = ImageFolderConfig
+
+    IMAGE_EXTENSIONS: List[str] = []  # definition at the bottom of the script
+
+    def _info(self):
+        return datasets.DatasetInfo(features=self.config.features)
+
+    def _split_generators(self, dl_manager):
+        if not self.config.data_files:
+            raise ValueError(f"At least one data file must be specified, but got data_files={self.config.data_files}")
+
+        labels = set()
+
+        def capture_labels_for_split(files, downloaded_files):
+            for file, downloaded_file in zip(files, downloaded_files):
+                file, downloaded_file = str(file), str(downloaded_file)
+                if os.path.isfile(downloaded_file):
+                    _, file_ext = os.path.splitext(file)
+                    if file_ext in self.IMAGE_EXTENSIONS:
+                        labels.add(os.path.basename(os.path.dirname(file)))
+                else:
+                    for downloaded_dir_file in dl_manager.iter_files(downloaded_file):
+                        _, downloaded_dir_file_ext = os.path.splitext(downloaded_dir_file)
+                        if downloaded_dir_file_ext in self.IMAGE_EXTENSIONS:
+                            labels.add(os.path.basename(os.path.dirname(downloaded_dir_file)))
+
+        data_files = self.config.data_files
+        downloaded_data_files = dl_manager.download_and_extract(data_files)
```

Comment:
This is unexpectedly long when you have a folder of 100,000 images, it gets stuck for ~2min on "Extracting data files"

Reply:
Even if you set `ignore_verifications` to `True` in `load_dataset` ?

Reply:
Yes even with `ignore_verifications=True`

Reply:
I tried with a local directory containing around 100,000 images from here: http://cs231n.stanford.edu/tiny-imagenet-200.zip

Reply:
After discussing the issue via Slack, the problem seems to stem from `ExtractorManager`'s checks. We can optimize this part by using `dl_manager.download` on an image file and `dl_manager.download_and_extract` on archives. I'll address this in a separate PR.